pkg/storage/stores/shipper/storage/cached_client.go

```diff
@@ -0,0 +1,153 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log/level"
+
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
+
+	"github.com/grafana/loki/pkg/storage/chunk"
+)
+
+const (
+	cacheTimeout = 1 * time.Minute
+)
+
+type table struct {
+	commonObjects []chunk.StorageObject
+	userIDs       []chunk.StorageCommonPrefix
+	userObjects   map[string][]chunk.StorageObject
+}
+
+type cachedObjectClient struct {
+	chunk.ObjectClient
+
+	tables       map[string]*table
+	tableNames   []chunk.StorageCommonPrefix
+	tablesMtx    sync.RWMutex
+	cacheBuiltAt time.Time
+
+	rebuildCacheChan chan struct{}
+	err              error
+}
+
+func newCachedObjectClient(downstreamClient chunk.ObjectClient) *cachedObjectClient {
+	return &cachedObjectClient{
+		ObjectClient:     downstreamClient,
+		tables:           map[string]*table{},
+		rebuildCacheChan: make(chan struct{}, 1),
+	}
+}
+
+func (c *cachedObjectClient) List(ctx context.Context, prefix, _ string) ([]chunk.StorageObject, []chunk.StorageCommonPrefix, error) {
+	prefix = strings.TrimSuffix(prefix, delimiter)
+	ss := strings.Split(prefix, delimiter)
+	if len(ss) > 2 {
+		return nil, nil, fmt.Errorf("invalid prefix %s", prefix)
+	}
+
+	if time.Since(c.cacheBuiltAt) >= cacheTimeout {
+		// when the cache is expired, only one concurrent call must be able to rebuild it
+		// all other calls will wait until the cache is built successfully or failed with an error
+		select {
+		case c.rebuildCacheChan <- struct{}{}:
+			c.err = nil
+			c.err = c.buildCache(ctx)
+			<-c.rebuildCacheChan
+			if c.err != nil {
+				level.Error(util_log.Logger).Log("msg", "failed to build cache", "err", c.err)
+			}
+		default:
+			for time.Since(c.cacheBuiltAt) >= cacheTimeout && c.err == nil {
+				time.Sleep(time.Millisecond)
+			}
+		}
+	}
+
+	if c.err != nil {
+		return nil, nil, c.err
+	}
+
+	c.tablesMtx.RLock()
+	defer c.tablesMtx.RUnlock()
+
+	if prefix == "" {
+		return []chunk.StorageObject{}, c.tableNames, nil
+	} else if len(ss) == 1 {
+		tableName := ss[0]
+		table, ok := c.tables[tableName]
+		if !ok {
+			return []chunk.StorageObject{}, []chunk.StorageCommonPrefix{}, nil
+		}
+
+		return table.commonObjects, table.userIDs, nil
+	} else {
+		tableName := ss[0]
```

Comment:
this else include the possibility of `len(ss) == 0` I think we should add a `if len(ss) >1 {`

Reply:
The length would never be `0` since `strings.Split` always returns at least one element in the list even if the separator is not there in the string. If you do `strings.Split("", "/")`, it would return `[]string{""}`